controller.py

```diff
@@ -460,231 +408,34 @@ def register(self, do_get=HTTP.debuggable_get):
         #)
         if isinstance(integration_contact_email, ProblemDetail):
             return integration_contact_email
-        hyperlinks_to_create = []
+
+        registrar = LibraryRegistrar(self._db, do_get=do_get)
+        result = registrar.register(
+            auth_url, shared_secret, library_stage
+        )
+        if isinstance(result, ProblemDetail):
+            return result
+        (library, library_is_new, from_shared_secret, auth_document,
+         hyperlinks_to_create) = result
```

Comment:
The main downside of this branch is the ugliness of the return value from `LibraryRegistrar.register`. There's a bunch of stuff that's determined during the course of registration, but which is only relevant to the controller. I'd like to hear your thoughts on how to improve this. The only thing I can think of is making a special `RegistrationStatus` class to contain this information, which isn't much better.

Reply:
I think the controller could be responsible for looking up or creating the library, so the registrar doesn't have to deal with the shared secret. I think it would be okay for that to happen before attempting to get the auth document, since it can get rolled back later if necessary.